src/libfsm/print/llvm.c

```diff
@@ -0,0 +1,824 @@
+		assert("unreached");
+		return NULL;
+	}
+}
+
+static void
+print_decl(FILE *f, const char *name, unsigned n)
+{
+	fprintf(f, "\t%%%s%u = ", name, n);
+}
+
+static void
+vprint_label(FILE *f, bool decl, const char *fmt, va_list ap)
+{
+	assert(fmt != NULL);
+
+	if (!decl) {
+		fprintf(f, "label %%");
+	}
+
+	vfprintf(f, fmt, ap);
+
+	if (decl) {
+		fprintf(f, ":\n");
+	}
+}
+
+static void
+print_label(FILE *f, bool decl, const char *fmt, ...)
+{
+	va_list ap;
+
+	va_start(ap, fmt);
+	vprint_label(f, decl, fmt, ap);
+	va_end(ap);
+}
+
+static void
+print_target(FILE *f, const struct dfavm_op_ir *op,
+	const char *fmt, ...)
+{
+	va_list ap;
+
+	if (op == &fail) {
+		print_label(f, false, "fail");
+	} else if (op != NULL && op->instr == VM_OP_STOP && op->cmp == VM_CMP_ALWAYS && op->u.stop.end_bits == VM_END_FAIL) {
+		print_label(f, false, "fail");
+	} else if (op != NULL) {
+		print_label(f, false, "l%" PRIu32, op->index);
+	} else if (fmt == NULL) {
+		assert("!unreached");
+		abort();
+	} else {
+		va_start(ap, fmt);
+		vprint_label(f, false, fmt, ap);
+		va_end(ap);
+	}
+}
+
+static void
+print_cond(FILE *f, const struct dfavm_op_ir *op, const struct fsm_options *opt,
+	struct frame *frame)
+{
+	assert(frame != NULL);
+	assert(op->cmp != VM_CMP_ALWAYS);
+
+	print_decl(f, "r", decl(&frame->r));
+	fprintf(f, "icmp %s i8 %%c%u, ",
+		cmp_operator(op->cmp), use(&frame->c));
+	llvm_escputcharlit(f, opt, op->cmp_arg);
+	fprintf(f, " ; ");
+	c_escputcharlit(f, opt, op->cmp_arg); // C escaping for a comment
+	fprintf(f, "\n");
+}
+
+static void
+print_jump(FILE *f, const struct dfavm_op_ir *dest)
+{
+	fprintf(f, "\tbr ");
+	print_target(f, dest, NULL);
+	fprintf(f, "\n");
+}
+
+static void
+print_incr(FILE *f, const char *name, unsigned n)
+{
+	// TODO: emit incr()?
+	print_decl(f, "n.new", n);
+	fprintf(f, "add i32 1, %%%s%u\n", name, n);
+
+	fprintf(f, "\tstore i32 %%%s.new%u, %s %%%s\n", name, n, ptr_i32, name);
+}
+
+static void
+print_branch(FILE *f, const struct frame *frame,
+	const struct dfavm_op_ir *op_true, const struct dfavm_op_ir *op_false)
+{
+	fprintf(f, "\tbr i1 %%r%u, ", use(&frame->r));
+	print_target(f, op_true, "t%u", use(&frame->b));
+	fprintf(f, ", ");
+	print_target(f, op_false, "f%u", use(&frame->b));
+	fprintf(f, "\n");
+}
+
+static void
+print_fetch(FILE *f, const struct fsm_options *opt,
+	struct frame *frame, const struct dfavm_op_ir *next,
+	enum dfavm_op_end end_bits)
+{
+	unsigned n = decl(&frame->c);
+	unsigned b = decl(&frame->b);
+
+	assert(frame != NULL);
+	assert(next != NULL);
+
+	/*
+	 * Emitting phi here would mean we'd need to keep track of the
+	 * basic blocks that jump to the current block. That's entirely
+	 * possible with our IR, but seems like a lot of work when we
+	 * can just alloca and let llvm do it for us.
+	 */
+
+	// TODO: consider emitting fetch(%s, %n)
+
+	switch (opt->io) {
+	case FSM_IO_GETC: {
+		print_decl(f, "i", n);
+		fprintf(f, "tail call i32 %%fsm_getc(%s noundef %%getc_opaque) #1\n",
```

Comment:
As far as I understand, it's a note about the ABI for the function you call, and it not needing to push a stack frame. That makes sense to be used for TCO returns, but apparently it can be used for other situations too!

I'm not sure if it's the right thing here, it's entirely possible it's wrong.

Reply:
(I'm happy to find out whenever I come to use it)